src/ocl/tensorocl.cpp

```diff
@@ -2146,7 +2148,9 @@ void CastTensor(const Handle& handle,
 
         const std::vector<std::size_t>& lens = srcDesc_flat.GetLengths();
 
-        std::string network_config = "cast " + std::to_string(dstDesc_flat.GetType());
+        // TODO: make proper network config
+        std::string network_config = "cast " + std::to_string(srcDesc_flat.GetType()) +
+                                     std::to_string(dstDesc_flat.GetType());
```

Comment:
Was this wrong before?

Looks like it used to have a caching key based off the dst type only.

Reply:
Before it did not use `srcType` for the network config case and failed if you do call `int` -> `float` and then `float` -> `int` conversion. In the both cases it calls `int` -> `float` because it has been added first.

Reply:
Btw - that's one of the "gtest single binary" problems - we may stuck withing the first added kernel.

Reply:
Proper fix is to include all the compile-time and performance critical parameters into the network config.
In the first case it will choose a properly compiled kernel and in the second one it will choose the most performant one (if there are any). I haven't analyzed it deep and just fixed that part with has been found we the wider coverage.